src/renderer/kube-object.store.ts

```diff
@@ -267,31 +269,68 @@ export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemSt
   }
 
   subscribe(apis = this.getSubscribeApis()) {
-    let disposers: {(): void}[] = [];
-
-    const callback = (data: IKubeWatchEvent) => {
-      if (!this.isLoaded) return;
-
-      this.eventsBuffer.push(data);
-    };
+    const abortController = new AbortController();
+    const namespaces = [...this.loadedNamespaces];
 
-    if (this.context.cluster?.isGlobalWatchEnabled && this.loadedNamespaces.length === 0) {
-      disposers = apis.map(api => api.watch({
-        namespace: "",
-        callback: (data) => callback(data),
-      }));
+    if (this.context.cluster?.isGlobalWatchEnabled && namespaces.length === 0) {
+      apis.forEach(api => this.watchNamespace(api, "", abortController));
     } else {
-      apis.map(api => {
+      apis.forEach(api => {
         this.loadedNamespaces.forEach((namespace) => {
-          disposers.push(api.watch({
-            namespace,
-            callback: (data) => callback(data)
-          }));
+          this.watchNamespace(api, namespace, abortController);
         });
       });
     }
 
-    return () => disposers.forEach(dispose => dispose());
+    return () => {
+      abortController.abort();
+    };
+  }
+
+  private watchNamespace(api: KubeApi<T>, namespace: string, abortController: AbortController) {
+    let timedRetry: NodeJS.Timeout;
+
+    abortController.signal.addEventListener("abort", () => clearTimeout(timedRetry));
+
+    const callback = (data: IKubeWatchEvent, error: any) => {
+      if (!this.isLoaded || abortController.signal.aborted) return;
+
+      if (error instanceof Response) {
+        if (error.status === 404) {
+          // api has gone, let's not retry
+          return;
+        } else { // not sure what to do, best to retry
+          timedRetry = setTimeout(() => {
+            api.watch({
```

Comment:
`api.watch()` returns a disposer for `unwatch: () => void` but it's not collected anywhere here?
yes, it could be aborted/disposed from above i guess, but it's not obvious

Reply:
First I collected all disposers but then I realised that it's duplicate work because we pass `abortController` reference to watches. When abort happens it will clean things.